specs/data_structures.md

```diff
@@ -475,7 +481,101 @@ enum VoteType : uint8_t {
+| `commissionRate`  | [Decimal](#decimal)         | Commission rate.                                                                       |
+| `delegatedCount`  | `uint32`                    | Number of accounts delegating to this validator.                                       |
+| `votingPower`     | `uint64`                    | Total voting power as staked balance + delegated stake, in `4u`.                       |
+| `pendingRewards`  | `uint64`                    | Rewards collected so far this period, in `1u`.                                         |
+| `latestEntry`     | [PeriodEntry](#periodentry) | Latest entry, used for calculating reward distribution.                                |
+| `unbondingHeight` | `uint64`                    | Block height validator began unbonding.                                                |
+| `isSlashed`       | `bool`                      | If this validator has been slashed or not.                                             |
+| `slashRate`       | [Decimal](#decimal)         | _Optional_, only if `isSlashed` is set. Rate at which this validator has been slashed. |
+
+Validator objects represent all the information needed to be keep track of a validator. Validators have four statuses:
+1. `Queued`: This validator has entered the queue to become an active validator. Once the next validator set transition occurs, if this validator has sufficient voting power (including its own stake and stake delegated to it) to be in the top `MAX_VALIDATORS` validators by voting power, it will become an active, i.e. `Bonded` validator. Until bonded, this validator can immediately exit the queue.
+1. `Bonded`: This validator is active and bonded. It can propose new blocks and vote on proposed blocks. Once bonded, an active validator must go through an unbonding process until its stake can be freed.
+1. `Unbonding`: This validator is in the process of unbonding, which can be voluntary (the validator decided to stop being an active validator) or forced (the validator committed a slashable offence and was kicked from the active validator set). Validators will remain in this status for at least `UNBONDING_DURATION` blocks, and while unbonding may still be slashed.
+1. `Unbonded`: This validator has completed its unbonding and has withdrawn its stake. The validator object will remain in this status until `delegatedCount` reaches zero, at which point it is destroyed.
+
+In the validators subtrees, validators are keyed by the [hash](#hashdigest) of their [address](#address). The first byte is then replaced with `ACTIVE_VALIDATORS_SUBTREE_ID` for the active validator set or `INACTIVE_VALIDATORS_SUBTREE_ID` for the inactive validator set. Active validators are `Bonded`, while inactive validators are not `Bonded`. By construction, the validators subtrees will be a subset of a mirror of the [accounts subtree](#account).
```

Comment:
I don't really understand why we would need two trees here. Or why we need a tree for the inactive validator set.

Reply:
It's purely an optimization. Every time a delegation operation (delegating, undelegating) is made  on an inactive validator, an a validator moving to/from inactive and active, you have to remove the validator object from the active validator tree to somewhere else. If it's the accounts, then you have to do all those operations on the entire accounts subtree (larger tree and no parallelization). If it's in a dedicated inactive validators subtree, then that subtree will be smaller and computing the roots can be done in parallel trivially.

Not strictly necessary but more efficient and easier to implement in a more performant way. Remember that we need to serialize the state (i.e. compute the new state root) after every transaction for fraud proofs.

Reply:
Thanks for the explanantion!